$\qquad$ $\qquad$$\qquad$  **TDA 231 Machine Learning: Homework 2** <br />
$\qquad$ $\qquad$$\qquad$ **Goal: Classification**<br />
$\qquad$ $\qquad$$\qquad$                   **Grader: Divya** <br />
$\qquad$ $\qquad$$\qquad$                     **Due Date: 23/4** <br />
$\qquad$ $\qquad$$\qquad$                   **Submitted by: Oskar Hulthén, 950801-1195, huoskar@student.chalmers** <br />
$\qquad$ $\qquad$$\qquad$                  ** And: Alexander Branzell, 931003-1977, alebra@student.chalmers.se** <br />

General guidelines:
* All solutions to theoretical problems, can be submitted as a single file named *report.pdf*. They can also be submitted in this ipynb notebook, but equations wherever required, should be formatted using LaTeX math-mode.
* All discussion regarding practical problems, along with solutions and plots should be specified here itself. We will not generate the solutions/plots again by running your code.
* Your name, personal number and email address should be specified above and also in your file *report.pdf*.
* All datasets can be downloaded from the course website.
* All tables and other additional information should be included.

# Theoretical problems

## [Naive Bayes Classifier, 6 points]

A psychologist does a small survey on ''happiness''. Each respondent provides a vector with entries 1 or 0 corresponding to if they answered “yes” or “no” to a question respectively. The question vector has attributes 
$$
x = (\mbox{rich, married, healthy}) \tag{1}
$$

Thus a response $(1, 0, 1)$ would indicate that the respondent was
''rich'', ''unmarried'' and ''healthy''. In addition, each respondent
gives a value $c = 1$ if they are content wih their life and $c = 0$
if they’re not. The following responses were obtained.

$$
c = 1: (1, 1, 1),(0, 0, 1),(1, 1, 0),(1, 0, 1) \\
c = 0: (0, 0, 0),(1, 0, 0),(0, 0, 1),(0, 1, 0)
$$

1. Using naive Bayes, what is the probability that a person is ''not rich'', ''married'' and ''healthy'' is ''content''?

2. What is the probability that a person who is ''not rich'' and ''married'' is content ? (i.e. we do not know if they are ''healthy'')

## [Extending Naive Bayes, 4 points]

Consider now, the following vector of attributes:

* $x_1 = 1$ if customer is younger than 20 and 0 otherwise.
* $x_2 = 1$ if customer is between 20 and 30 in age, and 0 otherwise.
* $x_3 = 1$ if customer is older than 30 and 0 otherwise
* $x_4 = 1$ if customer walks to work and 0 otherwise.

Each vector of attributes has a label ''rich'' or ''poor''. Point out potential difficulties with your approach above to training using naive Bayes. Suggest and describe how to extend your naive Bayes method to this dataset.


# Answers
For answers to the theoretical questions see the pdf report attached together with this notebook file on hand-in.

# Practical problems

## [Bayes classifier, 5 points]

Dowload the dataset **"dataset2.txt"**. You can use the following code for example:
```python
from numpy import genfromtxt
data = genfromtxt('dataset2.txt', delimiter=',')
labels = data[:,-1]
```
The dataset contains $3$-dimensional data, $X$, generated from $2$ classes with labels, $y$ either $+1$ or $-1$.  Each row of $X$ and $y$ contain one observation and one label respectively.  There are $1000$ instances of each class. 

a. Assume that the class conditional density is spherical Gaussian, and both classes have equal prior. Write the expression for the Bayes (<span style="color:red"> not **naive Bayes**</span>) classifier i.e. derive
$$
P(y_{new} = -1 | x_{new} , X, y ) \\
P(y_{new} = +1 | x_{new} , X, y ) ~.
$$

It is useful to note that the dependence on training data $X, y$ for class $1$ can be expressed as: 

$$ 
P( x_{new} | y_{new} = 1, X, y) = P(x_{new} |
\hat{\mu}_{1}, \hat{\sigma}^{2}_{1})
$$

where $\hat{\mu}_{1} \in \mathbb{R}^3$ and $\hat{\sigma}^{2}_{1}\in \mathbb{R}$ are MLE estimates for mean (3-dimensional) and variance based on training data with label $+1$ (and similarly for class 2 with label $-1$). 

b. Implement a function **sph_bayes()** which computes the probability of a new test point *Xtest* coming from class $1$ ($P1$) and class $2$ ($P2$). Finally, assign a label *Ytest* to the test point based on the probabilities $P1$ and $P2$.

```python
def sph_bayes(Xtest, ...): # other parameters needed.

    return [P1, P2, Ytest]
```
c. Write a function **new_classifier()**

```python
def new_classifier(Xtest, mu1, mu2)
    
    return [Ytest]
```
which implements the following classifier,
$$
f(x) = \mbox{sign}\left(\frac{(\mu_1 - \mu_2)^\top (x - b) }{\|\mu_1 -  \mu_2\|_2} \right)
$$
with $b = \frac{1}{2}(\mu_1 + \mu_2)$.

d. Report 5-fold cross validation error for both classifiers.


# Answers:
## a)
Following the Bayes formula for prediction:
$$P(t_{new}=k\ |\ x_{new},\mathbf{X},\mathbf{t}) = \frac{p(x_{new}\ |\ t_{new}=k,\mathbf{X},\mathbf{t})P(t_{new}=k)}{\sum _{j}p(x_{new}\ |\ t_{new}=j,\mathbf{X},\mathbf{t})P(t_{new}=j)}$$

<br>
<br>
Where we have two classes $y_{new}=-1$ and $y_{new}=+1$. To predict the +1 class we get the following:

$$P(y_{new}=+1\ |\ x_{new},\mathbf{X},\mathbf{t}) = \frac{p(x_{new}\ |\ y_{new}=+1,\mathbf{X},\mathbf{t})P(y_{new}=+1)}{p(x_{new}\ |\ y_{new}=+1,\mathbf{X},\mathbf{t})P(y_{new}=+1) + p(x_{new}\ |\ y_{new}=-1,\mathbf{X},\mathbf{t})P(y_{new}=-1)}$$
<br>
<br>
Given in the task is that both classes have equal prior distribution ($P(y_{new}=-1) = P(y_{new}=+1)$), meaning that the prior distribution can be removed:
<br>

$$P(y_{new}=+1\ |\ x_{new},\mathbf{X},\mathbf{t}) = \frac{p(x_{new}\ |\ y_{new}=+1,\mathbf{X},\mathbf{t})}{p(x_{new}\ |\ y_{new}=+1,\mathbf{X},\mathbf{t}) + p(x_{new}\ |\ y_{new}=-1,\mathbf{X},\mathbf{t})}$$
<br>
<br>


Furthermore we're given the way to express dependence upon the training data as $P( x_{new}\ |\ y_{new} = 1, X, y) = P(x_{new} | \hat{\mu}_{1}, \hat{\sigma}^{2}_{1})$ yielding the following final expressions for Bayes classifier:
<br>

$$P(y_{new}=+1\ |\ x_{new},\mathbf{X},\mathbf{t}) = \frac{P(x_{new}\ |\ \hat{\mu}_{+1}, \hat{\sigma}^{2}_{+1})}{P(x_{new}\ |\ \hat{\mu}_{+1}, \hat{\sigma}^{2}_{+1}) + P(x_{new}\ |\ \hat{\mu}_{-1}, \hat{\sigma}^{2}_{-1})}$$

<br>
<br>
$$P(y_{new}=-1\ |\ x_{new},\mathbf{X},\mathbf{t}) = \frac{P(x_{new}\ |\ \hat{\mu}_{-1}, \hat{\sigma}^{2}_{-1})}{P(x_{new}\ |\ \hat{\mu}_{+1}, \hat{\sigma}^{2}_{+1}) + P(x_{new}\ |\ \hat{\mu}_{-1}, \hat{\sigma}^{2}_{-1})}$$



In [227]:
# Task b),c) and d)
from numpy import genfromtxt
import numpy as np
import random
import matplotlib.pyplot as plt
from sklearn import datasets
from sklearn import model_selection

# Load the data
data = genfromtxt('dataset2.txt', delimiter=',')
labels = data[:,-1]

# Returns the Most likely estimator for mu and sigma based on the given data and class
# dimensions is the number of dimensions upon the data 
def mle_values(data, classifier, dimensions=3):
    n=0
    
    sum_mu  = np.zeros(dimensions) # The same amount of mu values as dimensions
    ## Calculate the mu value for the data with class == classifier
    for row in data:
        if(row[dimensions] == classifier):  # The class of the data is located in the column after the data
            for i in range (0,dimensions):  
                sum_mu[i] += row[i]         # add the data to the sum
            n += 1
    mu = sum_mu / float(n)            # Divide the sum by the amount of points to get the mu values
    
    sum_sig = 0
    ## Calculate the sigma value for the data with class == classifier
    for row in data:
        if(row[dimensions] == classifier):
            sum_sig += np.dot(np.transpose(row[0:3] - mu), row[0:3]-mu)  
    sig = np.sqrt(sum_sig) / float(n)
       
    # Return the MLE values for the given data with the given classifier
    return mu, sig

# Uses the multivariate gaussian pdf to return the probability for a given point with the sigma / mu
def sph_gauss(Xtest, mu, sig, dimensions=3):
    term1 = (1/ np.sqrt( (2* np.pi)**dimensions * np.sqrt(dimensions)*sig**2))
    term2 = np.exp(-1 / (2*sig**2) * np.sum(np.dot(np.transpose(Xtest-mu),Xtest-mu)**2) )
    return term1*term2

# Predicts a class upon the input Xtest based on the given mus and sigmas
def sph_bayes(Xtest, mu_1, sig_1 , mu_2 , sig_2):
    
    #Calculate the probability for each of the classes (for themselves)
    term1 = sph_gauss(Xtest,mu_1, sig_1)
    term2 = sph_gauss(Xtest,mu_2, sig_2)
    # Tie breaker: (The point seem to be precisely on the decision boundary)
    if (term1 == 0 and term2 == 0):
            term1 = random.random()
            term2 = 1-term1
    # Find the actual probability for each of the classes
    P1 = term1 / (term1+term2)
    P2 = term2 / (term1+ term2)
    
    # Pick a random value and "pick" based on the probabilites
    r = random.random()
    if (r < P1):
        Ytest = 1
    else:
        Ytest = -1
        
    # Returns the probabilites for each of the classes and the label that was predicted
    return [P1, P2, Ytest]

# Predicts a classifier upon the input Xtest based on the given mus
def new_classifier(Xtest, mu1, mu2, dimensions=3):
    # Follows the formula given in the task
    mu = mu1 - mu2
    point = Xtest- 1/2 * (mu1+mu2)
    dist = np.sqrt(np.sum((mu)**2))
    Ytest =np.sign( np.dot(np.transpose(mu), point) / dist)  
    
    return Ytest

# Main
def taskd():
    print("====== 5-fold cross validation ======")
    print("\n")
    KF = model_selection.KFold(n_splits=5, shuffle=True)
    x = 1
    error_sph = np.zeros(5)
    error_new = np.zeros(5)
    # Splits the data 5 times into training and test data
    for train_index, test_index in KF.split(data):
        print("==== Iteration %i ====" % x)
        
        # Get the mle values for each of the classes for the training data
        mu_1, sig_1 = mle_values(data[train_index], 1)
        mu_m1, sig_m1 = mle_values(data[train_index], -1)
        
        bayes = 0
        new = 0
        n = 0
        # Loop through the test data to predict a class using both classifiers
        # Also compare the prediction with the actual class
        for i in test_index:
            n += 1
            [_,_,res1] = sph_bayes(data[i][0:3],mu_1, sig_1, mu_m1, sig_m1)
            if res1 == data[i][3]:
                bayes += 1
            res2 = new_classifier(data[i][0:3],mu_1 , mu_m1)
            if res2 == data[i][3]:
                new += 1
        # Present the errors for each iteration
        error_sph[x-1] = 100 *(1 - bayes/float(n))
        error_new[x-1] = 100 * (1 - new/float(n))
        print("Error of spherical bayes: %.2f %% " % error_sph[x-1] )
        print("Error of new classifier: %.2f %%" % error_new[x-1] )
        print("\n")
        x += 1
    print("====== 5-fold cross validation errors: ======")
    sphere_5_fold = sum(error_sph) / 5
    new_5_fold = sum(error_new) / 5 
    print("Average error of spherical bayes: %.2f %% " % sphere_5_fold )
    print("Average error of new classifier: %.2f %%" % new_5_fold )
                      
taskd()

====== 5-fold cross validation ======


==== Iteration 1 ====
Error of spherical bayes: 0.50 % 
Error of new classifier: 25.50 %


==== Iteration 2 ====
Error of spherical bayes: 0.75 % 
Error of new classifier: 23.75 %


==== Iteration 3 ====
Error of spherical bayes: 0.25 % 
Error of new classifier: 25.50 %


==== Iteration 4 ====
Error of spherical bayes: 0.25 % 
Error of new classifier: 25.00 %


==== Iteration 5 ====
Error of spherical bayes: 0.75 % 
Error of new classifier: 27.00 %


====== 5-fold cross validation errors: ======
Average error of spherical bayes: 0.50 % 
Average error of new classifier: 25.35 %


## [DIGITS dataset classifer, 5 points]

Load the DIGITS dataset:
```python
from sklearn import datasets
digits = datasets.load_digits()
```
This dataset contains $1797$ samples of ten handwritten digit classes. You can further query and visualize the dataset using the various attributes of the returned dictionary:
```python
data = digits.data
print(data.shape)
target_names = digits.target_names
print (target_names)
import matplotlib.pyplot as plt
y = digits.target
plt.matshow(digits.images[0])
plt.show()
```

a. Use **new_classifier()** designed previously to do binary classification between classes representing digits "*5*" and "*8*".

b. Investigate an alternative feature function as described below:

1. Scale each pixel value to range $[0, 1] $ from original gray-scale ($0-255$). 
2. Compute variance of each row and column of the image. This will give you a new feature vector of size $16$ i.e. 

$$ 
x' = \left[ \; Var(row_1)  , Var(row_2), \ldots , Var(row_{8}), Var(col_1), \ldots, Var(col_{8}) \;\right]^T
$$

c. Report $5$-fold cross validation results for parts $(a)$ and
$(b)$ in a single table. What can you say about the results?

In [241]:
from sklearn import datasets
import matplotlib.pyplot as plt
from numpy import genfromtxt
import numpy as np
import random
from sklearn import model_selection


digits = datasets.load_digits()

data = digits.data
target_names = digits.target_names


#Modified from question 3
def new_classifier(Xtest, mu1, mu2):
    d = len(mu1)
    
    mu = np.zeros(d)
    for i in range(0, d):
        mu[i] = mu1[i] - mu2[i]
        
    b = np.zeros(d)
    
    for i in range(0, d):
        b[i] = 0.5 * (mu1[i] + mu2[i])
    point = Xtest-b
    dist = np.sqrt(np.sum((mu)**2))

    Ytest =np.sign( np.dot(np.transpose(mu),point) / dist)

    return Ytest


####a.)
#This function gets the mean of each pixel value in two sets of images and returns it as a 64-float array
def get_mean(fives, eights):
    mu_fives = np.zeros(64)
    for i in range(0, len(fives[0])):
        mu_fives[i] = np.mean(fives[:][i])

    mu_eights = np.zeros(64)
                         
    for i in range(0, len(eights[0])):
        mu_eights[i] = np.mean(eights[:][i])

    return mu_fives, mu_eights


####b.)
#This function takes two sets of images with pixel values between 0-16 and normalizes them to 0-1,
#then finds the variance in each row and column.
def get_normalized_variance(fives, eights):
    fives = np.asarray(fives)
    eights = np.asarray(eights)
    fives = fives/16.0
    eights = eights/16.0
    
    #first 8 values are row variance, last 8 values are column variance
    sigma_fives = np.zeros(16)
    sigma_eights = np.zeros(16)

    #Find the feature vector for all images in fives.
    for i in range(0, len(fives[0])):
        sigma_fives += get_feature_vector(fives[i])

    sigma_fives = sigma_fives / len(fives[0])
    
    for i in range(0, len(eights[0])):
        sigma_eights += get_feature_vector(eights[i])

    sigma_eights = sigma_eights / len(eights[0])
    return sigma_fives, sigma_eights

#Takes an array with 64 values, splits in into a matrix with 8x8 columns and finds
#the variance over each row and each column, which is returned in an array with 16 elements.
#The first 8 elements are the row variances, while the last are the column variances.
def get_feature_vector(data):
    tmp = np.split(data, 8)
    tmp = np.array(tmp)
    x = np.zeros(16)
    x[:8] += tmp.var(axis = 1)
    x[8:] += tmp.var(axis = 0)
    return x

####c.)
fives = data[digits.target == 5]
eights = data[digits.target == 8]

#Create a model to split the data with, shuffle the values so we
#won't get so many consecutive fives and eights.
KF = model_selection.KFold(n_splits=5, shuffle=True)
k = 1

#The total values to be used later
total_a = 0.0
total_b = 0.0
#Merge the fives and eights data
data = np.asarray(fives.tolist() + eights.tolist())

#Split the data into training and testing data based on the model.and start iterating.
for train_index, test_index in KF.split(data):
    training_fives = []
    training_eights = []

    #check every training index.
    for i in train_index:
        #Check if the index in the training data is a five or an eight.
        #The weird boolean is used to compare lists.
        if((fives == data[i]).all(1).any()):
            training_fives.append(data[i])
        else:
            training_eights.append(data[i])
    
    print("==== Iteration %i ====" % k)
    k += 1
    #from a.)
    mu_5, mu_8 = get_mean(training_fives, training_eights)
    #from b.)
    feature_5, feature_8 = get_normalized_variance(training_fives, training_eights)
    
    task_a = 0.0
    task_b = 0.0
    n = 0
    for i in test_index:
        n += 1
        a = new_classifier(data[i], mu_5, mu_8)
        #Check f we managed to classify a correct five with the method used in a.)
        if ((fives == data[i]).all(1).any() and a ==1):
            task_a += 1
        #Else, check if it was a correctly classified eight
        elif ((eights == data[i]).all(1).any() and a == -1):
            task_a += 1
        #Get the feature vector for the image
        x = get_feature_vector(data[i])
        b = new_classifier(x, feature_5, feature_8)
        #Check f we managed to classify a correct five with the method used in b.)
        if((fives == data[i]).all(1).any() and b == 1):
            task_b +=1
        #Else, check if it was a correctly classified eight
        elif ((eights == data[i]).all(1).any() and b == -1):
            task_b += 1
    
    print("Accuracy of task a.): %.3f %%" %(100 * task_a/float(n)) )
    print("Accuracy of task b.): %.3f %%" %(100 * task_b/float(n)) )
    print("\n")
    total_a += (task_a/float(n))
    total_b += (task_b/float(n))

total_a = total_a/5
total_b = total_b/5
print("==== Average accuracy a.) = %.3f %%" % (total_a*100))
print("==== Average accuracy b.) = %.3f %%" % (total_b*100))



    
    

###

==== Iteration 1 ====
Accuracy of task a.): 36.111 %
Accuracy of task b.): 69.444 %


==== Iteration 2 ====
Accuracy of task a.): 64.789 %
Accuracy of task b.): 67.606 %


==== Iteration 3 ====
Accuracy of task a.): 70.423 %
Accuracy of task b.): 67.606 %


==== Iteration 4 ====
Accuracy of task a.): 49.296 %
Accuracy of task b.): 87.324 %


==== Iteration 5 ====
Accuracy of task a.): 42.254 %
Accuracy of task b.): 76.056 %


==== Average accuracy a.) = 52.574 %
==== Average accuracy b.) = 73.607 %


# Observations
From the results we can see that the classifier using features is producing more reliable results than the classifier using means. The classifier using features usually does so with an accuracy of 70-80%, while the accuracy of the mean classifier can shift from 30% to 90% per iteration. This is because the classifier in task a can't distinguish between fives or eights very well, and will continuously guess that most images are fives. This means that when a subset of the data containing mostly fives is selected, the mean classifier performs ''well'', while when the subset is containing mostly eights it performs poorly. This also means that the classifier using features is more stable since it gives a similar performance regardless of what subset of the data it is given.

|     | 1  | 2  | 3  | 4  | 5  | Average |
|-----|----|----|----|----|----|---------|
| a.) | 36.11% | 64.79% | 70.42% | 49.30% | 42.25% | 52.57%      |
| b.) | 69.44% | 67.61% | 67.61% | 67.60% | 87.32% | 73.61%      |